flooder/src/main.rs

```diff
@@ -0,0 +1,365 @@
+    let total_threads = config.threads;
+    let total_batches = config.transactions / config.throughput;
+    let users_per_thread = total_users / total_threads;
+    let transactions_per_batch = config.throughput / config.threads;
+    let transfer_amount = 1u128;
+
+    info!(
+        "Using account {} to derive and fund accounts",
+        &account.public()
+    );
+
+    let users_and_nonces /*(users, nonces) */= derive_user_accounts(
+        connection.clone(),
+        account.clone(),
+        total_users,
+        transfer_amount,
+        total_batches,
+    )
+    .await;
+
+    debug!("all accounts have received funds");
+
+    let txs = sign_transactions(
+        connection.clone(),
+        account,
+        users_and_nonces,
+        transfer_amount,
+        total_threads,
+        total_batches,
+        users_per_thread,
+    )
+    .await;
+
+    println!(
+        "Prepared {} signed txs",
+        txs.len() * txs[0].len() * txs[0][0].len()
+    );
+
+    let histogram = Arc::new(Mutex::new(
+        HdrHistogram::<u64>::new_with_bounds(1, u64::max_value(), 3).unwrap(),
+    ));
+
+    println!(
+        "threads: {}, batches: {}, txs per batch: {}",
+        &txs.len(),
+        &txs[0].len(),
+        &txs[0][0].len()
+    );
+
+    let tick = Instant::now();
+
+    flood(
+        pool,
+        txs,
+        total_threads,
+        total_batches,
+        transactions_per_batch,
+        &histogram,
+    )
+    .await;
+
+    let tock = tick.elapsed().as_millis();
+    let histogram = histogram.lock().unwrap();
+
+    println!("Summary:\n TransferTransactions sent: {}\n Total time:        {} ms\n Slowest tx:        {} ms\n Fastest tx:        {} ms\n Average:           {:.1} ms\n Throughput:        {:.1} tx/s",
+             histogram.len (),
+             tock,
+             histogram.max (),
+             histogram.min (),
+             histogram.mean (),
+             1000.0 * histogram.len () as f64 / tock as f64
+    );
+
+    Ok(())
+}
+
+async fn flood(
+    pool: Vec<Api<sr25519::Pair, WsRpcClient>>,
+    txs: Vec<Vec<Vec<TransferTransaction>>>,
+    total_threads: u64,
+    total_batches: u64,
+    transactions_per_batch: u64,
+    histogram: &Arc<Mutex<HdrHistogram<u64>>>,
+) {
+    // send txs
+    for batch in 0..total_batches {
+        let mut futures_batch = Vec::new();
+
+        println!("Preparing batch {}", batch);
+
+        for thread in 0..total_threads {
+            println!("sending batch {} from thread {}", batch, thread);
+
+            for tx in 0..transactions_per_batch {
+                // retrieve signed txs
+                let signed_tx = &txs[thread as usize][batch as usize][tx as usize];
+
+                // send txs using a connection pool
+                futures_batch.push(send_tx(
+                    pool.get(tx as usize % pool.len()).unwrap().to_owned(),
+                    signed_tx.to_owned(),
+                    Arc::clone(histogram),
+                ));
+            }
```

Comment:
I think one future/thread should get more txs to send at once. So the `for tx in 0..transactions_per_batch` I would put into the future code.

Reply:
refactored, see if you like it more this way. Found a great lib (rayon, https://github.com/rayon-rs/rayon) that is an abstraction over async/await and provides sugar for parallel iterators